drivers/tee/tee_core.c

```diff
@@ -302,6 +307,60 @@ static int params_from_user(struct tee_context *ctx, struct tee_param *params,
 	return 0;
 }
 
+static int params_from_user_ocall(struct tee_context *ctx,
+				  struct tee_param *params, size_t num_params,
+				  struct tee_ioctl_param __user *uparams)
+{
+	/* For comments, see params_from_user */
+
+	size_t n;
+
+	for (n = 0; n < num_params; n++) {
+		struct tee_shm *shm;
+		struct tee_ioctl_param ip;
+
+		if (copy_from_user(&ip, uparams + n, sizeof(ip)))
+			return -EFAULT;
+
+		/* All unused attribute bits have to be zero */
+		if (ip.attr & ~TEE_IOCTL_PARAM_ATTR_MASK)
+			return -EINVAL;
+
+		params[n].attr = ip.attr;
+		switch (ip.attr & TEE_IOCTL_PARAM_ATTR_TYPE_MASK) {
+		case TEE_IOCTL_PARAM_ATTR_TYPE_NONE:
+			break;
+		case TEE_IOCTL_PARAM_ATTR_TYPE_VALUE_INPUT:
+		case TEE_IOCTL_PARAM_ATTR_TYPE_VALUE_INOUT:
+		case TEE_IOCTL_PARAM_ATTR_TYPE_VALUE_OUTPUT:
```

Comment:
`TEE_IOCTL_PARAM_ATTR_TYPE_VALUE_OUTPUT` doesn't need to copy values.

Reply:
Consider what `OUTPUT` means in the context of an OCALL though.

If the current invocation of the ECALL IOCTL is a function invocation (i.e., not a reply to an OCALL), then you are right, the TA is the one that fills these values.

If the current invocation of the ECALL IOCTL is a reply to an OCALL, though, `OUTPUT` works in reverse: it is filled by the CA.

So what I could do is distinguish between these two cases and copy `INPUT` and `INOUT` values when the current invocation is not an OCALL reply, and copy `INOUT` and `OUTPUT` values when the current invocation is an OCALL reply.